File: app/src/main/java/com/github/qezt/phonenumberidentifier/CallReceiver.java
package com.github.qezt.phonenumberidentifier;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.provider.ContactsContract;
import android.telephony.TelephonyManager;
import android.widget.Toast;

import com.android.volley.Response;
import com.android.volley.VolleyError;

public class CallReceiver extends BroadcastReceiver {
    @Override
    public void onReceive(Context context, Intent intent) {
        String state = intent.getStringExtra(TelephonyManager.EXTRA_STATE);
        String incomingNumber = intent.getStringExtra(TelephonyManager.EXTRA_INCOMING_NUMBER);

        if (state == null || ! state.equals(TelephonyManager.EXTRA_STATE_RINGING)) {
            return;
        }

        Uri uri = Uri.withAppendedPath(ContactsContract.PhoneLookup.CONTENT_FILTER_URI, Uri.encode(incomingNumber));
        Cursor cursor = Application.instance().getContentResolver().query(uri, new String[]{ContactsContract.PhoneLookup.DISPLAY_NAME}, null, null, null);
        if (cursor != null) {
            try {
                if (cursor.getCount() > 0) return;
            } finally {
                cursor.close();
            }
        }

        RequestManager.instance().addRequest(new QueryPhoneIdRequest(incomingNumber, new PhoneIdentityListener(), new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                Toast.makeText(Application.instance(), "Bummer", Toast.LENGTH_SHORT).show();
            }
        }));
    }
}

File: app/src/main/java/com/github/qezt/phonenumberidentifier/CallReceiver.kt
package com.github.qezt.phonenumberidentifier

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.ContactsContract
import android.telephony.TelephonyManager
import android.widget.Toast

class CallReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {

        val state = intent.getStringExtra(TelephonyManager.EXTRA_STATE)
        val incomingNumber = intent.getStringExtra(TelephonyManager.EXTRA_INCOMING_NUMBER)
        CallReceiver.state = state
        if (state != TelephonyManager.EXTRA_STATE_RINGING) {
            return
        }
        if (incomingNumber == null) {
            Toast.makeText(Application.app, "Number is empty", Toast.LENGTH_LONG).show()
            return
        }
        Toast.makeText(Application.app, "Number is $incomingNumber", Toast.LENGTH_LONG).show()

        val uri = Uri.withAppendedPath(ContactsContract.PhoneLookup.CONTENT_FILTER_URI, Uri.encode(incomingNumber))
        val cursor = Application.app.contentResolver.query(uri, arrayOf(ContactsContract.PhoneLookup.DISPLAY_NAME), null, null, null)
        if (cursor != null) {
            try {
                if (cursor.count > 0) return
            } finally {
                cursor.close()
            }
        }

        Coroutine.launchOnUI {
            val result = Coroutine.asyncOnWorkers {
                QueryPhoneIdRequest.get(incomingNumber)
            }.await()
            if (result == null) {
                Toast.makeText(Application.app, "Bummer", Toast.LENGTH_SHORT).show()
                return@launchOnUI
            }
            PhoneIdentityPopup.show(result)
        }
    }

    companion object {
        private var state: String? = null
        fun duringCall(): Boolean {
            return state == null || state != TelephonyManager.EXTRA_STATE_IDLE
        }
    }
}

